### PR TITLE
Documentation for the CoreEval feature

### PR DESCRIPTION
**Pull request by Tyrosine22**
Here's the initial pass at CoreEval documentation. 

BTW, some of the changes which originally appeared in this PR have been moved to the revisedgettingstarted PR (https://github.com/Agoric/documentation/pull/811).

The gist with the output from the initial set of console commands is located here: https://gist.github.com/Tyrosine22/6da2367ae8897736953c2f1286164eed

The gist with the output from the final set of console commands is located here:
https://gist.github.com/Tyrosine22/12ba3222f15926f25eee30ce86be33f2

### CODE REVIEW

File: main/guides/coreeval/permissions.md
Comment (ivanlei) [inline main/guides/coreeval/permissions.md:66]: grqnts -> grants

File: main/guides/coreeval/permissions.md
Comment (ivanlei) [inline main/guides/coreeval/permissions.md:67]: _fix_:
We rarely refer to ERTP assets as "money". Perhaps "mint ERTP assets"?

File: main/guides/coreeval/permissions.md
Comment (ivanlei) [inline main/guides/coreeval/permissions.md:69]: _fix_:
the vote -> a vote

File: main/guides/coreeval/proposal.md
Comment (ivanlei) [inline main/guides/coreeval/proposal.md:3]: _question_:
The script below doesn't make use of `storageNode`. What is it meant to highlight or explain?

File: main/guides/coreeval/permissions.md
Comment (ivanlei) [inline main/guides/coreeval/permissions.md:88]: _question_:
How do the `brand section` and `issuer section` differ?

File: main/guides/coreeval/permissions.md
Comment (ivanlei) [inline main/guides/coreeval/permissions.md:3]: _comment_:
The example is somewhat out of date with the code on master. Probably not a big deal.

**Review by ivanlei — CHANGES_REQUESTED** [review]
Can you please make requested changes and reply to questions. 
How were the example shell commands tested? Can you add to the PR description a block of console output showing each example shell command being run and the successful output of them?

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:66]: Fixed.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:67]: Right you are. Fixed.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:69]: Fixed.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:88]: I added the following sentence to the Brand Section to clarify: "Note that any brands that the contract may mint are not included in this section, unless they're also produced or consumed in some other capacity."

File: main/guides/coreeval/proposal.md
Comment (Tyrosine22) [reply main/guides/coreeval/proposal.md]: I rewrote the text blurbs at the top of both this page & the README.md page. Take a look at both, please.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:3]: I definitely think that's fine. The additional changes are contract-specific. However, one of the changes appears to be upgrade-related, so after I document the upgrade API, I want to return to this section, update the permissions page, and discuss the additional properties.

Comment (dckc) [inline main/guides/coreeval/permissions.md:3]:
```suggestion
Write a json file with the permissions that the proposal will need. For example, [gov-add-psm-permit.json](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/test/psm/gov-add-psm-permit.json) is the proposal Agoric created for the PSM contract:
```

Please let's keep accessibility in mind and use [meaningful link text](https://www.w3.org/WAI/tips/writing/#make-link-text-meaningful), avoiding "here" links.

File: main/guides/coreeval/permissions.md
Comment (dckc) [inline main/guides/coreeval/permissions.md:63]: `bankManager` is closely related to ERTP (and to cosmos) but not to Zoe, so I suggest taking Zoe out.

And I think most contracts will not need this very powerful capability: It's the right to manipulate nearly all cosmos assets in the chain.

That is: it's the capability/right to make a "bank" for any address, where a "bank" is a per-address right to make an ERTP purse for any denom registered in the vbank... which raises the question of "what's the vbank?" so we might want to leave this level of detail out of scope.

File: main/guides/coreeval/README.md
Comment (michaelfig) [inline main/guides/coreeval/README.md:3]: I took another stab at clarifying the last sentence, please consider with a grain of salt:

```suggestion
Agoric facilitates safely injecting code to the Agoric testnet, usually for the purpose of running a contract and modifying it based on the result of a governance vote. To do so, we submit a proposal using cosmos-sdk level API ([swingset.CoreEval](https://community.agoric.com/t/blder-dao-governance-using-arbitrary-code-injection-swingset-coreeval/99)). If the proposal passes its governance vote, its JS code is run with ocaps extracted using the proposal's permit list, which the code can combine to perform privileged tasks.
```

File: main/guides/coreeval/permissions.md
Comment (michaelfig) [inline main/guides/coreeval/permissions.md:67]:
```suggestion
* **feeMintAccess**: When this permission is set to "zoe", the contract will be able to create digital assets. Only contracts that mint privileged Agoric digital assets (not the unprivileged [`zcf.makeMint`](some link here)) will need access to this.
```

File: main/guides/coreeval/permissions.md
Comment (michaelfig) [inline main/guides/coreeval/permissions.md:66]: It's probably worth mentioning above here that to "set" a permission, its value can be any truthy primitive.  The value itself is not meaningful to the proposal (`true`, `"zoe"`, `"timer"`, `"razzleflatz"` and `249`, are interchangeable and mean the same thing).

As a footnote, Agoric conventionally names its permission values for the identity of the home vat of the ocap corresponding to the permission.  This can be useful for visualizing the produce/consume dependencies between actions.

Then, I'd change all the "permission is set to XXX" just to be "permission is set".

```suggestion
* **zoe**: When this permission is set, it grants access to the Zoe framework. All contracts will need access to this.
```

**Review by michaelfig — COMMENTED** [review]
Looking good so far!  I have some suggestions, and will defer approval to another reviewer.

**Comment by dckc** [comment]
oops... I should have packaged up my comments into a reivew, I guess. But I thought I was just going to take a peek... then I saw 1 thing... then 1 more...

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:3]: Fixed.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:63]: The description for bankManager has been changed to: 

"Grants access to bank-related functionality within ERTP, allowing the contract to manipulate nearly all Cosmos assets in the chain. Because this capability is very powerful, this permission should only be granted to contracts that absolutely need it."

File: main/guides/coreeval/README.md
Comment (Tyrosine22) [reply main/guides/coreeval/README.md:3]: The sentence has been changed to your suggestion.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:67]: feeMintAccess's description has been changed to pretty much match your suggested text.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:66]: An explanation was added to the top of the Permissions page saying that any string will set a permission. Similarly, all the permission descriptions were simplified per your suggestion above.

File: main/guides/coreeval/permissions.md
Comment (dckc) [reply main/guides/coreeval/permissions.md:93]: doesn't look like the change made it to this PR yet.

If you're talking about a change you haven't pushed yet, fair enough; I'll stand by.

Comment (michaelfig) [reply main/guides/coreeval/permissions.md:93]: This looks fixed now.

**Comment by Tyrosine22** [comment]
> Can you please make requested changes and reply to questions.
> How were the example shell commands tested? Can you add to the PR description a block of console output showing each example shell command being run and the successful output of them?

Here's the gist with the output: https://gist.github.com/Tyrosine22/6da2367ae8897736953c2f1286164eed

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [inline main/guides/coreeval/local-testnet.md:23]:
```suggestion
6. Navigate to `<agoric-sdk>/bin` and submit the governance proposal.
```

It doesn't start a client running.

Comment (dckc) [reply main/guides/coreeval/local-testnet.md:23]: And this isn't part of starting a local chain; it shouldn't be under than heading.

File: main/guides/coreeval/proposal.md
Comment (dckc) [inline main/guides/coreeval/proposal.md:6]: The `startPSM` global was provided in the pismo boot configuration but isn't provided in vaults.

The global environment for CoreEval scripts is another thing we should document:


https://github.com/Agoric/agoric-sdk/blob/bf164a05f425460ad357841b537b8dbc044357a2/packages/vats/src/core/chain-behaviors.js#L78-L83


where `farExports` comes from `import * as farExports from '@endo/far';` and hence includes `E` and `Far`.
`endowments` comes from a few lines previous:

https://github.com/Agoric/agoric-sdk/blob/bf164a05f425460ad357841b537b8dbc044357a2/packages/vats/src/core/chain-behaviors.js#L40-L45

Documenting `VatData` is a whole other thing; maybe we can leave that out of scope for now.

And `allPowers.modules` comes from:

https://github.com/Agoric/agoric-sdk/blob/bf164a05f425460ad357841b537b8dbc044357a2/packages/vats/src/core/boot-chain.js#L17-L20

`startPSM` was part of `allPowers.modules` in the pismo bootstrap. But in vaults, it's not there.

File: main/guides/getting-started/README.md
Comment (dckc) [inline main/guides/getting-started/README.md:47]:
```suggestion
## Install jq
```

`jq` is just [jq](https://stedolan.github.io/jq/). jQuery is https://jquery.com/

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [inline main/guides/coreeval/local-testnet.md:6]:
```suggestion
1. Install *Go* if needed. For example:
```

Installing `go` with `brew` assumes a mac with `brew` installed. Our [platform section](https://docs.agoric.com/guides/getting-started/#platform-linux-shell-or-equivalent) says we support linux, MacOS, and WSL.

Building the `go` stuff is something we left out of the SDK 'till now. It merits its own section. In [Getting Started](https://docs.agoric.com/guides/getting-started/). The idea was that it wasn't needed for smart contract development. With the smart wallet architecture, this is no longer really viable.

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [inline main/guides/coreeval/local-testnet.md:20]: can we use `agoric start local-chain` instead? I'd rather not promote the `scenario2` stuff in the cosmic-swingset `Makefile` to a product feature.

File: main/guides/coreeval/permissions.md
Comment (dckc) [inline main/guides/coreeval/permissions.md:69]:
```suggestion
* **board**: Grants write access to the [board name service](https://docs.agoric.com/guides/wallet/#the-agoric-board).
```

That is: (1) the board isn't specific to the PSM and (2) we have a section that documents it; let's refer to it. (I might not be using the right URL convention)

File: main/guides/coreeval/permissions.md
Comment (dckc) [inline main/guides/coreeval/permissions.md:78]: I'd rather not promote `testFirstAnchorKit` to a documented product feature. It would be more typical to produce a brand or an issuer... and perhaps a `kit` holding the admin and creator facets like [feeDistributorKit](https://github.com/Agoric/agoric-sdk/blob/bf164a05f425460ad357841b537b8dbc044357a2/packages/inter-protocol/src/proposals/econ-behaviors.js#L50), `reserveKit`, `auctioneerKit` etc.

File: main/guides/coreeval/permissions.md
Comment (dckc) [reply main/guides/coreeval/permissions.md:93]: It still talks about minting brands.

Also, the brand section is all about sharing the brands created by a contract. I don't think "not included in this section" is on target.

The question was: how are the `brand` section and the `issuer` section different? They're different in that `makeIssuerKit` creates both a brand and an issuer (as well as a mint). The contract creator typically wants to share both of them with prospective clients. The brand is typically shared by producing an item in the `brand` section, and the issuer is shared by producing an item in the `issuer` section. The `permit` here is so that governance can control the namespace.

File: main/guides/getting-started/README.md
Comment (dckc) [inline main/guides/getting-started/README.md:12]: `brew` isn't available on linux.

Also, jQuery and jq are different things.

Comment (dckc) [reply main/guides/getting-started/README.md:12]: Please let's move any changes to the getting-started section to its own PR.

File: main/guides/getting-started/README.md
Comment (dckc) [inline main/guides/getting-started/README.md:31]: Again, let's move any changes to the getting-started section, such as changing the node version, to its own PR.

**Review by dckc — CHANGES_REQUESTED** [review]
It's getting there, but there are still a number of corrections that are important.

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [inline main/guides/coreeval/local-testnet.md:1]: We already have a section for this: https://docs.agoric.com/guides/dapps/starting-multiuser-dapps.html

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:93]: I'm resolving this conversation here, and taking the discussion to Dan's official review below where the same issue is raised.

Comment (dckc) [reply main/guides/coreeval/permissions.md:93]: > I'm not seeing it mention minting brands anymore.

ah. I see the fix now. thanks.

Comment (dckc) [reply main/guides/coreeval/permissions.md:93]: > unless the Issuer is also named "DAI"?

Yes, as usual, the DAI asset has 3 facets; brand, issuer, and mint. The DAI brand is shared via the `brand` section; the DAI issuer is shared in the `issuer` section, and the DAI mint is held closely rather than being shared.

File: main/guides/coreeval/local-testnet.md
Comment (Tyrosine22) [reply main/guides/coreeval/local-testnet.md:23]: I made the changes suggested above.

In addition, to address your second issue, (i.e., "it shouldn't be under that heading") I change the title of the topic to "Deploy a Governance Proposal to a Local  Testnet" so that it properly reflects all the actions described in the topic.

File: main/guides/coreeval/proposal.md
Comment (Tyrosine22) [reply main/guides/coreeval/proposal.md:6]: It sounds like you're saying that startPSM isn't a part of CoreEval's global environment anymore. But gov-add-psm.js is checked-in code, it appears to assume that startPSM is part of its global environment, and the code works when I run it. So it seems like startPSM is still in global?

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:69]: Fixed. 

Also, although your URI was well-formed & valid, I converted it to the internal-reference format that the other URI in the docs use. (i.e. the URI is now /guides/wallet/README.md#the-agoric-board)

File: main/guides/coreeval/proposal.md
Comment (dckc) [reply main/guides/coreeval/proposal.md:6]: > `gov-add-psm.js` is checked-in code

yes, but there are no automated tests for it

> the code works when I run it.

That surprises me. There are no errors in the agd logs? And you can see a new PSM contract started with `agd query vstorage children published.psm.IST`?

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [reply main/guides/coreeval/local-testnet.md:6]: p.s. the https://docs.agoric.com/guides/dapps/starting-multiuser-dapps.html looks like it discusses building the `go` stuff.

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [reply main/guides/coreeval/local-testnet.md:23]: > I made the changes suggested above.

Did you push it yet? The most recent commit I see in this PR is on May 24 b22d2dd8d02b7b8b3bf6210d1d763eb4565687f8 .

File: main/guides/coreeval/proposal.md
Comment (dckc) [reply main/guides/coreeval/proposal.md:6]: @michaelfig , @Tyrosine22 says you were there when it worked for him. Any idea how the global reference to `startPSM` worked? I'm confused.

Comment (dckc) [reply main/guides/coreeval/proposal.md:6]: Suggestion: just say "The script executes in an environment with globals such as `E` and `Far` provided." and leave enumeration of the whole list of globals to a later PR.

Comment (michaelfig) [reply main/guides/coreeval/proposal.md:6]: > @michaelfig , @Tyrosine22 says you were there when it worked for him.

We verified that the vote passed and that something happened on the chain.

> Any idea how the global reference to `startPSM` worked? I'm confused.

I don't know if the proposal contents were viable or completed successfully, only that they were initiated.

File: main/guides/coreeval/local-testnet.md
Comment (Tyrosine22) [reply main/guides/coreeval/local-testnet.md:6]: I've moved the installation of Go to the Getting Started section in another PR.

Also, I changed the installation instructions so they're platform agnostic. (I hadn't realized that "brew" was Mac-specific.)

File: main/guides/getting-started/README.md
Comment (Tyrosine22) [reply main/guides/getting-started/README.md]: Fixed. (And moved to another PR)

File: main/guides/coreeval/permissions.md
Comment (dckc) [inline main/guides/coreeval/permissions.md:82]: Should we be explicit that "well-known" means that it goes into (or comes from) `agoricNames`?

That is: `installation: { consume: { contractGovernor: true } }` means that the script will get access to the equivalent of `E(agoricNames).lookup('installation', 'contractGovernor')`

Or perhaps leave the details of the NameHub API until later? I just made an issue for it:

 - #807

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:78]: I'm not promoting testFirstAnchorKit so much as using it as an example. It's the only item listed in PSM's Produce section, so I'm (briefly) mentioning what it is.

BTW, if a Brand were to be produced, wouldn't it appear in the Brand section rather than this one?

File: main/guides/getting-started/README.md
Comment (Tyrosine22) [reply main/guides/getting-started/README.md]: This has been moved to another PR. The installation instructions have also been made platorm-agnostic.

File: main/guides/getting-started/README.md
Comment (Tyrosine22) [reply main/guides/getting-started/README.md]: Done.

File: main/guides/coreeval/permissions.md
Comment (dckc) [reply main/guides/coreeval/permissions.md:78]: > if a Brand were to be produced, wouldn't it appear in the Brand section

yes.

File: main/guides/coreeval/local-testnet.md
Comment (Tyrosine22) [reply main/guides/coreeval/local-testnet.md]: We can, but let's do this in a separate PR; I was able to troubleshoot the "This project isn't in a folder created by agoric init" error, but I've run into a different error, and this one doesn't have a helpful error message. I'll ping you or mfig to work through it.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:6]: Changed the text per your suggestion.

File: main/guides/coreeval/README.md
Comment (Tyrosine22) [reply main/guides/coreeval/README.md:3]: Right... throughout this PR the reader is shown how to inject code into a local Testnet, but presumably they'd also be able to inject code into a remote Testnet.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md]: Good point. The titles of the Top Level Consume & Top Level Produce sections have been changed to explicitly state that they're the top level.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:82]: Yeah, let's handle this issue in a separate PR.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md]: No, there wasn't any real reason for the differences. They were just terms carried over from different sources. The terms *should* be standardized, but I wasn't sure which term best captures what's going on. I'm still not sure, actually.

I've changed the topic to largely conform to your "capabilities" language. The terms should be more standardized (permit list vs. permissions. vs capabilities), but (as I said above) I'm not 100% sure about which term I like best, and the topic reads OK as-is. So... it's good enough or now, I suppose.

File: main/guides/coreeval/permissions.md
Comment (dckc) [inline main/guides/coreeval/permissions.md:96]: I still don't understand why this "not included in this section" note is here. It seems wrong.

**Review by dckc — APPROVED** [review]
This conveys the gist of what folks need to know, so I'm approving.

I would prefer that we finish 2 things before landing this, but they can probably be addressed later (or maybe even never) without much harm: The note about brands not being included doesn't seem right to me. Nor does the `startPSM` global.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md]: Fixed,

File: main/guides/coreeval/local-testnet.md
Comment (dckc) [inline main/guides/coreeval/local-testnet.md:33]: "How can I tell?" I can imagine our audience asking. I suggest providing an example log entry showing a block being produced.

(suggestion; not critical / important)

**Review by dckc — APPROVED** [review]
still looks good

File: main/guides/coreeval/proposal.md
Comment (Tyrosine22) [reply main/guides/coreeval/proposal.md:11]: The following text was added to the top blurb on proposal.md:

"The script executes in an environment with globals such as E and Far provided."

File: main/guides/coreeval/local-testnet.md
Comment (Tyrosine22) [reply main/guides/coreeval/local-testnet.md]: All the "make scenario2" commands have been converted to various "agoric XXX" commands.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md]: The text has been changed to the suggested passage.

File: main/guides/coreeval/permissions.md
Comment (Tyrosine22) [reply main/guides/coreeval/permissions.md:98]: It was based on a misunderstanding I had of what the Brand & Issuer sections contained. I've deleted the note.

**Comment by michaelfig** [comment]
@ivanlei, have your requested changes been addressed?

**Review by michaelfig — APPROVED** [review]

**Review by ivanlei — APPROVED** [review]
🚀

**Pull request merged**
